jupyterhub/app.py

```diff
@@ -698,6 +698,60 @@ def _bind_url_default(self):
         proto = 'https' if self.ssl_cert else 'http'
         return proto + '://:8000'
 
+    public_url = Unicode(
+        "",
+        config=True,
+        help="""Set the public URL of JupyterHub
+
+        This will skip any detection of URL and protocol from requests,
+        which isn't always correct when JupyterHub is behind
+        multiple layers of proxies, etc.
+        Usually the failure is detecting http when it's really https.
+
+        Should include the full, public URL of JupyterHub,
+        including the public-facing base_url prefix
+        (i.e. it should include a trailing slash).
+        """,
+    )
+
+    @default("public_url")
+    def _default_public_url(self):
+        if self.subdomain_host:
+            # if subdomain_host is specified, use it by default
+            return self.subdomain_host + self.base_url
+        else:
+            return ""
+
+    @validate("public_url")
+    def _validate_public_url(self, proposal):
+        url = proposal.value
+        if not url:
+            # explicitly empty (default)
+            return url
+        if not url.endswith("/"):
+            # ensure we have a trailing slash
+            # for consistency with base_url
+            url = url + "/"
+        if not url.endswith(self.base_url):
+            if not urlparse(url).path.strip("/"):
+                # no path specified, add base_url and warn
```

Comment:
Since  this is a URL prefix by definition, I think adding the slash is the right thing to do. If it were a canonical URL for a single page, that would be different, but it is a prefix so adding a slash is for our internal convenience and consistency when joining with sub-paths.

Reply:
Adding the slash is fine, it's the addition of `base_url` I was wondering about. Not a blocker though.

Reply:
Ah, gotcha. Yeah, that seems okay to raise. What we really care about here is the public _origin_ (proto://host:port), as base_url config already exists and is defined elsewhere, so specifying just the origin seems like sensible behavior as opposed to having to change base_url in multiple places. However, if you might be _hoping_ that public_url allows specifying a public URL that _doesn't_ match base_url would allow proxying, that's not going to work because we use absolute paths with base_url (you can't do e.g. proxy_pass with rewriting paths by adding or removing a prefix, for example), in which case an error could be helpful.